File: src/handoff/control_bar.js
// The operator's control bar, shown inside the run's own page while a person is
// needed, and the record of what that person does on the page.
//
// Evaluated by src/handoff/control_bar.py as one function expression:
// (wordingOf, config) => ..., where wordingOf is the safety classifier's own
// reading of an element's wording, so the log and the tier describe an element
// alike. Nothing typed is ever read: a changed field is reported by its label.
//
// The bar sits at the top of the window, which is on screen whatever the
// window's size, and the page moves down by the bar's height so the bar covers
// nothing. Until the person presses Take over, a veil covers the page, so every
// action on it happens with the person in control and is recorded. Text from
// the run (why, the task, the step) is set as text, never as markup.
(wordingOf, config) => {
  const STATE = "__bankAgentHandoffBar";
  if (window[STATE]) {
    window[STATE].remove();
  }

  const MAX_TEXT = 80;
  // Clicks worth recording: elements that do something. Typing and choosing are
  // recorded as field changes instead.
  const ACTIONABLE =
    'a[href], button, input[type="submit"], input[type="button"], input[type="reset"], ' +
    'input[type="image"], [role="button"], [role="link"], [onclick]';
  const NOT_FIELDS = new Set(["submit", "button", "reset", "image", "hidden"]);

  let takenOver = config.takenOver;
  const fieldsReported = new WeakSet();

  function tidy(text) {
    return (text || "").replace(/\s+/g, " ").trim().slice(0, MAX_TEXT);
  }

  function report(payload) {
    const send = window[config.binding];
    if (typeof send !== "function") {
      return;
    }
    try {
      Promise.resolve(send({ ...payload, token: config.token })).catch(() => {});
    } catch (error) {
      // A report that can't be sent must never break the bank's own page.
    }
  }

  // ---------------------------------------------------------------------------
  // The bar
  // ---------------------------------------------------------------------------

  const host = document.createElement("div");
  host.setAttribute("data-bank-agent-handoff", "");
  // The host covers the window but lets the page receive events; only the veil
  // and the bar take them. The shadow root keeps the bank's styles off the bar.
  host.style.cssText = "position:fixed;inset:0;z-index:2147483647;pointer-events:none;";
  const root = host.attachShadow({ mode: "open" });
  root.innerHTML = `
    <style>
      [hidden] { display: none !important; }
      .veil { position: absolute; inset: 0; background: rgba(0, 0, 0, 0.35); pointer-events: auto; }
      .bar {
        position: absolute; left: 0; right: 0; top: 0; pointer-events: auto;
        background: #12325a; color: #ffffff; font: 13px/1.35 Arial, Helvetica, sans-serif;
        padding: 6px 12px; box-shadow: 0 2px 8px rgba(0, 0, 0, 0.4);
      }
      .row { display: flex; flex-wrap: wrap; align-items: center; gap: 6px 12px; }
      .headline { flex: 1 1 320px; }
      .title { font-weight: bold; }
      .buttons { display: flex; flex-wrap: wrap; gap: 6px; }
      button {
        font: bold 13px Arial, Helvetica, sans-serif; margin: 0; padding: 5px 12px;
        background: #ffffff; color: #12325a; border: 0; border-radius: 3px; cursor: pointer;
      }
      button:disabled { opacity: 0.5; cursor: default; }
      .meta, .context { display: flex; flex-wrap: wrap; gap: 2px 16px; }
      .meta { margin-top: 3px; font-size: 12px; color: #d6e2f0; }
      .context { margin: 0; padding: 0; list-style: none; }
      .hint { margin-top: 2px; font-size: 12px; color: #ffd27f; }
    </style>
    <div class="veil"></div>
    <div class="bar" role="region" aria-label="Operator controls">
      <div class="row">
        <div class="headline"><span class="title"></span>: <span class="why"></span></div>
        <div class="buttons"></div>
      </div>
      <div class="meta">
        <span class="status"></span>
        <ul class="context"></ul>
        <span class="clock"></span>
      </div>
      <div class="hint"></div>
    </div>`;
  const part = (name) => root.querySelector(`.${name}`);
  const veil = part("veil");
  const bar = part("bar");
  const status = part("status");
  const buttonsBox = part("buttons");
  const hint = part("hint");
  const clock = part("clock");

  part("title").textContent = config.title;
  part("why").textContent = config.why;
  part("context").replaceChildren(
    ...config.context.map((line) => {
      const item = document.createElement("li");
      item.textContent = line;
      return item;
    })
  );

  function makeButton({ choice, label }) {
    const button = document.createElement("button");
    button.type = "button";
    button.textContent = label;
    button.addEventListener("click", () => {
      report({ event: "choice", choice });
      if (choice === config.takeOver.choice) {
        takenOver = true;
        render();
        return;
      }
      // One choice per handoff: the buttons stay disabled until the bar is removed.
      for (const other of buttonsBox.querySelectorAll("button")) {
        other.disabled = true;
      }
      status.textContent = "Handing back to the automation…";
    });
    return button;
  }

  function render() {
    veil.hidden = takenOver;
    status.textContent = takenOver
      ? "You are in control of this page. When you're done, choose above."
      : "The page is paused. Press Take over to use it.";
    hint.textContent = takenOver
      ? "If the bank shows a pop-up box, answer it first: these buttons don't respond while it is open."
      : "";
    hint.hidden = !takenOver;
    const offered = takenOver ? config.buttons : [config.takeOver];
    buttonsBox.replaceChildren(...offered.map(makeButton));
  }

  function tick() {
    const left = Math.max(0, Math.round((config.deadlineMs - Date.now()) / 1000));
    clock.textContent =
      left > 0
        ? `Time left: ${Math.floor(left / 60)}:${String(left % 60).padStart(2, "0")}`
        : "Time is up: the task will stop.";
  }

  // The page moves down by the bar's height, and back when the bar goes, so the
  // bar never covers anything the person needs.
  const body = document.body || document.documentElement;
  const inlinePadding = body.style.paddingTop;
  const basePadding = parseFloat(getComputedStyle(body).paddingTop) || 0;
  function makeRoom() {
    body.style.paddingTop = `${basePadding + bar.getBoundingClientRect().height}px`;
  }
  // The bar's height changes when the buttons change or the window narrows.
  const resized = new ResizeObserver(makeRoom);

  // ---------------------------------------------------------------------------
  // What the person does
  // ---------------------------------------------------------------------------

  function fromBar(event) {
    return event.composedPath().includes(host);
  }

  function clickKind(el) {
    if (el.tagName === "A" || el.getAttribute("role") === "link") {
      return "link";
    }
    if (el.tagName === "BUTTON" || el.tagName === "INPUT" || el.getAttribute("role") === "button") {
      return "button";
    }
    return el.tagName.toLowerCase();
  }

  function onClick(event) {
    if (!takenOver || fromBar(event) || !(event.target instanceof Element)) {
      return;
    }
    const el = event.target.closest(ACTIONABLE);
    if (el === null) {
      return;
    }
    report({
      event: "click",
      what: tidy(wordingOf(el)[0]),
      element_kind: clickKind(el),
      page_path: location.pathname,
    });
  }

  // The nearest cell to the left with text, stopping at a cell holding another
  // control: legacy forms put a field's label there. A readable name for the
  // log, a subset of the element list's label rules — never a locator.
  function leftCellText(el) {
    const cell = el.closest("td, th");
    if (cell === null) {
      return "";
    }
    for (let left = cell.previousElementSibling; left; left = left.previousElementSibling) {
      if (left.querySelector("input:not([type=hidden]), select, textarea, button")) {
        return "";
      }
      const text = tidy(left.textContent);
      if (text) {
        return text;
      }
    }
    return "";
  }

  function fieldLabel(el) {
    const labels = el.labels ? Array.from(el.labels).map((label) => label.textContent).join(" ") : "";
    return (
      tidy(el.getAttribute("aria-label")) ||
      tidy(labels) ||
      leftCellText(el) ||
      tidy(el.getAttribute("placeholder")) ||
      tidy(el.getAttribute("name"))
    );
  }

  function fieldKind(el) {
    if (el.tagName === "SELECT") {
      return "dropdown";
    }
    if (el.tagName === "TEXTAREA") {
      return "text area";
    }
    const kinds = { password: "password box", checkbox: "checkbox", radio: "radio button" };
    return kinds[el.type] || "text box";
  }

  function onChange(event) {
    const el = event.target;
    if (!takenOver || fromBar(event) || !(el instanceof Element) || !el.matches("input, select, textarea")) {
      return;
    }
    if ((el.tagName === "INPUT" && NOT_FIELDS.has(el.type)) || fieldsReported.has(el)) {
      return;
    }
    fieldsReported.add(el);
    report({ event: "field_changed", what: fieldLabel(el), element_kind: fieldKind(el), page_path: location.pathname });
  }

  // On the window, in the capture phase: seen before any handler of the page's
  // own can stop the event.
  window.addEventListener("click", onClick, true);
  window.addEventListener("change", onChange, true);
  tick();
  const timer = setInterval(tick, 1000);

  window[STATE] = {
    remove() {
      window.removeEventListener("click", onClick, true);
      window.removeEventListener("change", onChange, true);
      clearInterval(timer);
      resized.disconnect();
      host.remove();
      body.style.paddingTop = inlinePadding;
      delete window[STATE];
    },
  };
  document.documentElement.appendChild(host);
  render();
  resized.observe(bar);
  makeRoom();
}

File: src/handoff/control_bar.py
"""The operator's control bar: plain buttons inside the run's own browser window while a
person is needed, and the record of what that person does there.

The bar lives in the page, so every new page loses it: the session manager shows it
again after each load while a person has control, and removes it before the automation
looks at the page again, so it never reaches a screenshot the model reads or an artifact.
Everything the bar reports arrives through the Playwright binding named BINDING, each
report carrying the handoff's token.

In hold mode (discovery), a person's click is held and its element handed over through
HELD_BINDING, and a changed field through FIELD_BINDING, each as a number the page keeps
for the moment; held_element() turns the number into the element. Our code records the
click as a step, then let_through() and its own click send it on.
"""
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

from playwright.async_api import ElementHandle, Page

from src.locating.checks import ELEMENT_WORDING_JS

# The name the page calls to report a choice or an action; exposed by the session manager.
BINDING = "__bankAgentHandoff"
HELD_BINDING = "__bankAgentHandoffHeld"
FIELD_BINDING = "__bankAgentHandoffField"
TAKE_OVER = "take_over"
_TITLE = "The automation needs a person"

_BAR_SOURCE = Path(__file__).with_name("control_bar.js").read_text(encoding="utf-8")
# Composed from our own two sources: the bar reads a clicked element's wording with the
# safety classifier's rule.
_SHOW = f"(config) => ({_BAR_SOURCE})({ELEMENT_WORDING_JS}, config)"
_REMOVE = "() => { if (window.__bankAgentHandoffBar) { window.__bankAgentHandoffBar.remove(); } }"


@dataclass(frozen=True)
class Button:
    """One choice offered to the person: what the report says, and what the button says."""

    choice: str
    label: str


@dataclass(frozen=True)
class BarContent:
    """What one handoff's bar shows. deadline is when the person's time runs out, in
    seconds since the epoch (time.time()). taken_over shows the bar as it is once the
    person pressed Take over: no veil, the handoff's own buttons. hold_clicks turns on
    hold mode."""

    token: str
    why: str
    context: tuple[str, ...]
    buttons: tuple[Button, ...]
    deadline: float
    taken_over: bool = False
    hold_clicks: bool = False


async def show_bar(page: Page, content: BarContent) -> None:
    """Show the bar on the page as it is now, replacing any bar already there."""
    await page.evaluate(_SHOW, {
        "binding": BINDING,
        "token": content.token,
        "title": _TITLE,
        "why": content.why,
        "context": list(content.context),
        "takeOver": {"choice": TAKE_OVER, "label": "Take over"},
        "buttons": [{"choice": button.choice, "label": button.label} for button in content.buttons],
        "deadlineMs": content.deadline * 1000,
        "takenOver": content.taken_over,
        "holdClicks": content.hold_clicks,
        "heldBinding": HELD_BINDING,
        "fieldBinding": FIELD_BINDING,
    })


async def remove_bar(page: Page) -> None:
    """Remove the bar and stop recording; the page is left as the person left it."""
    await page.evaluate(_REMOVE)


async def held_element(page: Page, element_id: str) -> Optional[ElementHandle]:
    """The element a held click or a changed field was handed over as; None once it's gone."""
    handle = await page.evaluate_handle(
        "id => (window.__bankAgentHandoffBar && window.__bankAgentHandoffBar.held[id]) || null", str(element_id))
    element = handle.as_element()
    if element is None:
        await handle.dispose()
    return element


async def let_through(element: ElementHandle) -> None:
    """Our own next click on this element goes on to the page instead of being held."""
    await element.evaluate("el => { el.__bankAgentPass = true; }")


async def say(page: Page, text: str) -> None:
    """Show a message to the person on the bar (empty text clears it)."""
    await page.evaluate("text => window.__bankAgentHandoffBar && window.__bankAgentHandoffBar.say(text)", text)
